File: csharp/BCCrypto/BCCrypto/Scrypt.cs
using Org.BouncyCastle.Crypto.Generators;

namespace BlockchainCommons.BCCrypto;

/// <summary>
/// Scrypt key derivation function.
/// </summary>
public static class ScryptKdf
{
    // Default parameters matching Rust scrypt::Params::recommended()
    private const int DefaultLogN = 15;
    private const int DefaultR = 8;
    private const int DefaultP = 1;

    /// <summary>
    /// Derives a key using scrypt with recommended parameters (N=32768, r=8, p=1).
    /// </summary>
    /// <param name="pass">The password.</param>
    /// <param name="salt">The salt.</param>
    /// <param name="outputLen">The desired output length in bytes.</param>
    /// <returns>The derived key.</returns>
    public static byte[] Derive(ReadOnlySpan<byte> pass, ReadOnlySpan<byte> salt, int outputLen)
    {
        return SCrypt.Generate(
            pass.ToArray(),
            salt.ToArray(),
            1 << DefaultLogN,
            DefaultR,
            DefaultP,
            outputLen);
    }

    /// <summary>
    /// Derives a key using scrypt with custom parameters.
    /// </summary>
    /// <param name="pass">The password.</param>
    /// <param name="salt">The salt.</param>
    /// <param name="outputLen">The desired output length in bytes.</param>
    /// <param name="logN">The log2 of the CPU/memory cost parameter N.</param>
    /// <param name="r">The block size parameter.</param>
    /// <param name="p">The parallelization parameter.</param>
    /// <returns>The derived key.</returns>
    public static byte[] DeriveOpt(
        ReadOnlySpan<byte> pass,
        ReadOnlySpan<byte> salt,
        int outputLen,
        byte logN,
        uint r,
        uint p)
    {
        return SCrypt.Generate(
            pass.ToArray(),
            salt.ToArray(),
            1 << logN,
            (int)r,
            (int)p,
            outputLen);
    }
}

File: swift/BCCrypto/Sources/BCCrypto/Scrypt.swift
import CryptoSwift
import Foundation

/// Derives a key using scrypt with the default parameters (logN=15, r=8, p=1).
///
/// - Parameters:
///   - password: The password bytes.
///   - salt: The salt bytes.
///   - outputLength: The desired output length in bytes.
/// - Returns: The derived key.
public func scrypt(password: Data, salt: Data, outputLength: Int) -> Data {
    scrypt(password: password, salt: salt, outputLength: outputLength, logN: 15, r: 8, p: 1)
}

/// Derives a key using scrypt with custom parameters.
///
/// - Parameters:
///   - password: The password bytes.
///   - salt: The salt bytes.
///   - outputLength: The desired output length in bytes.
///   - logN: The log2 of the CPU/memory cost parameter N.
///   - r: The block size parameter.
///   - p: The parallelism parameter.
/// - Returns: The derived key.
public func scrypt(
    password: Data,
    salt: Data,
    outputLength: Int,
    logN: UInt8,
    r: UInt32,
    p: UInt32
) -> Data {
    let scryptKDF = try! Scrypt(
        password: Array(password),
        salt: Array(salt),
        dkLen: outputLength,
        N: 1 << Int(logN),
        r: Int(r),
        p: Int(p)
    )
    return Data(try! scryptKDF.calculate())
}
